File: Python3.x/101-Symmetric Tree.py
# 递归
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root: return True
        return self.judge(root.left, root.right)

    def judge(self, left, right):
        if not left or not right: return left == right
        if left.val != right.val: return False
        return self.judge(left.left, right.right) and self.judge(left.right, right.left)


'''
递归，对于每个节点，检查树的左右节点值是否相等，
同时判断：左节点的左子树和右节点的右子树是否对称、右节点的左子树和左节点的右子树是否对称。
'''

File: Algorithms/C++/101-Symmetric_Tree.cpp
// Runtime: 4 ms, faster than 83.66% of C++ online submissions for Symmetric Tree.
// Memory Usage: 14.7 MB, less than 93.22% of C++ online submissions for Symmetric Tree.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == NULL)
            return true;
        return judge(root->left, root->right);
    }
private:
    bool judge(TreeNode* left, TreeNode* right) {
        if (left == NULL || right == NULL)
            return left == right;
        if (left->val != right->val)
            return false;
        return judge(left->left, right->right) && judge(left->right, right->left);
    }
};
